cmd/src/search_stream.go

```diff
@@ -321,25 +317,44 @@ const streamingTemplate = `
 `
 
 var streamSearchTemplateFuncs = map[string]interface{}{
-	"streamSearchHighlightMatch": func(query string, match streaming.EventLineMatch) string {
-		var highlights []highlight
-		if strings.Contains(query, "patterntype:structural") {
-			highlights = streamConvertMatchToHighlights(match, false)
-			return applyHighlightsForFile(match.Line, highlights)
+	"streamSearchHighlightMatch": func(match streaming.ChunkMatch) string {
+		var result []rune
+
+		addLineColumn := func(n int) []rune {
+			result = append(result, []rune("  ")...)
+			result = append(result, []rune(ansiColors["search-line-numbers"])...)
+			result = append(result, []rune(fmt.Sprintf("% 6d", match.ContentStart.Line+1+n))...)
+			result = append(result, []rune(ansiColors["nc"])...)
+			result = append(result, []rune(ansiColors["search-border"])...)
+			result = append(result, []rune(" |  ")...)
+			result = append(result, []rune(ansiColors["nc"])...)
+			return result
 		}
 
-		highlights = streamConvertMatchToHighlights(match, true)
-		return applyHighlights(match.Line, highlights, ansiColors["search-match"], ansiColors["nc"])
-	},
+		lineCount := 0
+		addLineColumn(lineCount)
+		for offset, r := range match.Content {
+			offset++
+			if r == '\n' {
+				lineCount++
+				result = append(result, []rune(ansiColors["nc"])...)
+				result = append(result, r)
+				addLineColumn(lineCount)
+				result = append(result, []rune(ansiColors["search-match"])...)
+				continue
+			}
+			for _, rr := range match.Ranges {
+				if rr.Start.Offset-match.ContentStart.Offset+1 == offset {
+					result = append(result, []rune(ansiColors["search-match"])...)
+				} else if rr.End.Offset-match.ContentStart.Offset+1 == offset {
+					result = append(result, []rune(ansiColors["nc"])...)
+				}
+			}
```

Comment:
sticking to the style of highlighting logic we currently, which is to check whether the current rune is within any highlightable range. slicing the highlighted versus non-highlighted parts out of content felt not fun and I didn't want to spend that much time to this effort.

Reply:
Does this work with overlapping ranges? If I'm reading it correctly, we set a color if we find a range that starts at this rune, we unset the color if we find a range that ends at this rune. For an overlapping range, the end of the first range will cancel highlighted color early. I think this could be solved with just a "active ranges" counter. We practically only will have overlapping ranges in very few situations, but it's not impossible per the API definitions.